nydus-image does not compile on macos

See https://github.com/bergwolf/image-service/runs/6402632511?check_suite_focus=true
```
error[E0433]: failed to resolve: could not find `linux` in `os`
  --> src/bin/nydus-image/core/node.rs:13:14
   |
13 | use std::os::linux::fs::MetadataExt;
   |              ^^^^^ could not find `linux` in `os`

error[E0432]: unresolved import `nix::sys::stat::makedev`
  --> src/bin/nydus-image/builder/stargz.rs:17:5
   |
17 | use nix::sys::stat::makedev;
   |     ^^^^^^^^^^^^^^^^^^^^^^^ no `makedev` in `sys::stat`

error[E0405]: cannot find trait `MetadataExt` in this scope
   --> src/bin/nydus-image/core/node.rs:851:[35](https://github.com/bergwolf/image-service/runs/6402632511?check_suite_focus=true#step:4:35)
    |
851 |     fn meta(&self) -> Result<impl MetadataExt> {
    |                                   ^^^^^^^^^^^ not found in this scope
    |
help: consider importing one of these items
    |
8   | use crate::builder::diff::thread::thread::env::os_imp::MetadataExt;
    |
8   | use crate::builder::diff::thread::thread::env::sys::platform::fs::MetadataExt;
    |
8   | use std::os::macos::fs::MetadataExt;
    |
8   | use std::os::unix::prelude::MetadataExt;
    |

error[E0425]: cannot find function `major` in module `stat`
   --> src/bin/nydus-image/core/node.rs:986:41
    |
986 |         self.inode.is_chrdev() && stat::major(self.rdev) == 0 && stat::minor(self.rdev) == 0
    |                                         ^^^^^ not found in `stat`

error[E0425]: cannot find function `minor` in module `stat`
   --> src/bin/nydus-image/core/node.rs:986:72
    |
986 |         self.inode.is_chrdev() && stat::major(self.rdev) == 0 && stat::minor(self.rdev) == 0
    |                                                                        ^^^^^ not found in `stat`

error[E0[40](https://github.com/bergwolf/image-service/runs/6402632511?check_suite_focus=true#step:4:40)5]: cannot find trait `MetadataExt` in this scope
    --> src/bin/nydus-image/core/node.rs:1518:26
     |
1518 |     fn set_inode_info<T: MetadataExt>(
     |                          ^^^^^^^^^^^ not found in this scope
     |
help: consider importing one of these items
     |
8    | use crate::builder::diff::thread::thread::env::os_imp::MetadataExt;
     |
8    | use crate::builder::diff::thread::thread::env::sys::platform::fs::MetadataExt;
     |
8    | use std::os::macos::fs::MetadataExt;
     |
8    | use std::os::unix::prelude::MetadataExt;
     |

error: aborting due to 6 previous errors
```

--- a/.github/workflows/release.yml
+++ b/.github/workflows/release.yml
@@ -78,6 +78,8 @@ jobs:
         rustup target install $arch-apple-darwin
         make -e ARCH=$arch macos-fusedev
         sudo mv target-fusedev/$arch-apple-darwin/release/nydusd nydusd-fusedev
+        sudo mv target-fusedev/$arch-apple-darwin/release/nydus-cached .
+        sudo mv target-fusedev/$arch-apple-darwin/release/nydus-image .
         sudo mv target-fusedev/$arch-apple-darwin/release/nydusctl .
         sudo cp -r misc/configs .
         sudo chown -R $(id -un):$(id -gn) . ~/.cargo/

--- a/Makefile
+++ b/Makefile
@@ -92,8 +92,7 @@ ut:
 	RUST_BACKTRACE=1 ${CARGO} test $(VIRIOFS_COMMON) --bin nydusd -- --nocapture --test-threads=8
 
 macos-fusedev:
-	# nydus-cached/nydus-image does not support macos yet
-	${CARGO} build --target ${ARCH}-apple-darwin --target-dir ${current_dir}/target-fusedev --features=fusedev --release --bin nydusctl --bin nydusd
+	${CARGO} build --target ${ARCH}-apple-darwin --target-dir ${current_dir}/target-fusedev --features=fusedev --release --bin nydusctl --bin nydusd --bin nydus-image --bin nydus-cached
 
 macos-ut:
 	${CARGO} clippy --target-dir ${current_dir}/target-fusedev --features=fusedev --bin nydusd --release --workspace -- -Dwarnings

--- a/Makefile
+++ b/Makefile
@@ -3,7 +3,7 @@ all: build
 TEST_WORKDIR_PREFIX ?= "/tmp"
 DOCKER ?= "true"
 
-ARCH ?= $(shell uname -p)
+ARCH ?= $(shell uname -m)
 OS ?= linux
 CARGO ?= $(shell which cargo)
 CARGO_BUILD_GEARS = -v ~/.ssh/id_rsa:/root/.ssh/id_rsa -v ~/.cargo/git:/root/.cargo/git -v ~/.cargo/registry:/root/.cargo/registry

--- a/src/bin/nydus-cached/main.rs
+++ b/src/bin/nydus-cached/main.rs
@@ -12,20 +12,18 @@ extern crate nydus_storage as storage;
 
 use std::io::Result;
 use std::os::unix::prelude::AsRawFd;
-use std::sync::{atomic::AtomicBool, Mutex};
+use std::sync::atomic::AtomicBool;
 
 use clap::{App, Arg};
 use mio::unix::SourceFd;
 use mio::{Events, Interest, Poll, Token};
 use nydus_app::{dump_program_info, setup_logging, BuildTimeInfo};
 use nydus_storage::remote::{RemoteBlobMgr, Server};
-use vmm_sys_util::eventfd::EventFd;
 
 const CLIENT_TOKEN: Token = Token(1);
 
 lazy_static! {
     static ref EVENT_MANAGER_RUN: AtomicBool = AtomicBool::new(true);
-    static ref EXIT_EVTFD: Mutex::<Option<EventFd>> = Mutex::<Option<EventFd>>::default();
 }
 
 fn main() -> Result<()> {

--- a/src/bin/nydus-image/builder/diff.rs
+++ b/src/bin/nydus-image/builder/diff.rs
@@ -856,11 +856,12 @@ pub mod tests {
     use std::os::unix::fs as unix_fs;
     use std::path::Path;
 
-    use nix::sys::stat::{dev_t, makedev, mknod, Mode, SFlag};
+    use nix::sys::stat::{dev_t, mknod, Mode, SFlag};
     use nix::unistd::geteuid;
     use vmm_sys_util::tempdir::TempDir;
 
     use super::*;
+    use nydus_utils::compact::makedev;
     use nydus_utils::exec;
 
     fn create_dir(path: &Path) {

--- a/src/bin/nydus-image/builder/stargz.rs
+++ b/src/bin/nydus-image/builder/stargz.rs
@@ -191,7 +191,7 @@ impl TocEntry {
     }
 
     pub fn mode(&self) -> u32 {
-        let mut mode = self.mode;
+        let mut mode = 0;
 
         if self.is_dir() {
             mode |= libc::S_IFDIR;
@@ -207,7 +207,9 @@ impl TocEntry {
             mode |= libc::S_IFIFO;
         }
 
-        mode
+        #[cfg(target_os = "macos")]
+        let mode = mode as u32;
+        self.mode | mode
     }
 
     pub fn rdev(&self) -> u32 {

--- a/src/bin/nydus-image/builder/stargz.rs
+++ b/src/bin/nydus-image/builder/stargz.rs
@@ -14,9 +14,9 @@ use std::rc::Rc;
 use std::str::FromStr;
 
 use anyhow::{anyhow, bail, Context, Error, Result};
-use nix::sys::stat::makedev;
 use serde::{Deserialize, Serialize};
 
+use nydus_utils::compact::makedev;
 use nydus_utils::digest::{self, Algorithm, DigestHasher, RafsDigest};
 use nydus_utils::{try_round_up_4k, ByteSize};
 use rafs::metadata::layout::v5::{RafsV5ChunkInfo, RafsV5Inode, RafsV5InodeFlags};

--- a/src/bin/nydus-image/core/bootstrap.rs
+++ b/src/bin/nydus-image/core/bootstrap.rs
@@ -264,10 +264,13 @@ impl Bootstrap {
         }
 
         // dot & dotdot
-        node.dirents
-            .push((node.offset, OsString::from("."), libc::S_IFDIR));
-        node.dirents
-            .push((parent_offset, OsString::from(".."), libc::S_IFDIR));
+        #[allow(clippy::useless_conversion)]
+        {
+            node.dirents
+                .push((node.offset, OsString::from("."), libc::S_IFDIR.into()));
+            node.dirents
+                .push((parent_offset, OsString::from(".."), libc::S_IFDIR.into()));
+        }
 
         let mut dirs: Vec<&mut Tree> = Vec::new();
         for child in tree.children.iter_mut() {

--- a/src/bin/nydus-image/core/bootstrap.rs
+++ b/src/bin/nydus-image/core/bootstrap.rs
@@ -264,6 +264,9 @@ impl Bootstrap {
         }
 
         // dot & dotdot
+        // Type of libc::S_IFDIR is u16 on macos, so it need a conversion
+        // but compiler will report useless conversion on linux platform,
+        // so we add an allow annotation here.
         #[allow(clippy::useless_conversion)]
         {
             node.dirents

--- a/src/bin/nydus-image/core/node.rs
+++ b/src/bin/nydus-image/core/node.rs
@@ -1000,16 +1000,9 @@ impl Node {
         if spec != WhiteoutSpec::Overlayfs {
             return false;
         }
-
-        #[cfg(target_os = "linux")]
-        {
-            use nix::sys::stat;
-            self.inode.is_chrdev() && stat::major(self.rdev) == 0 && stat::minor(self.rdev) == 0
-        }
-        #[cfg(target_os = "macos")]
-        {
-            self.inode.is_chrdev() && ((self.rdev >> 24) & 0xff) == 0 && (self.rdev & 0xffffff) == 0
-        }
+        self.inode.is_chrdev()
+            && nydus_utils::compact::major_dev(self.rdev) == 0
+            && nydus_utils::compact::minor_dev(self.rdev) == 0
     }
 
     /// Check whether the inode (directory) is a overlayfs whiteout opaque.

--- a/src/bin/nydus-image/core/node.rs
+++ b/src/bin/nydus-image/core/node.rs
@@ -10,13 +10,15 @@ use std::fmt::{self, Display, Formatter, Result as FmtResult};
 use std::fs::{self, File};
 use std::io::{Read, SeekFrom, Write};
 use std::mem::size_of;
+#[cfg(target_os = "linux")]
 use std::os::linux::fs::MetadataExt;
+#[cfg(target_os = "macos")]
+use std::os::macos::fs::MetadataExt;
 use std::os::unix::ffi::OsStrExt;
 use std::path::{Component, Path, PathBuf};
 use std::str::FromStr;
 
 use anyhow::{Context, Error, Result};
-use nix::sys::stat;
 use sha2::digest::Digest;
 
 use nydus_utils::{
@@ -999,7 +1001,15 @@ impl Node {
             return false;
         }
 
-        self.inode.is_chrdev() && stat::major(self.rdev) == 0 && stat::minor(self.rdev) == 0
+        #[cfg(target_os = "linux")]
+        {
+            use nix::sys::stat;
+            self.inode.is_chrdev() && stat::major(self.rdev) == 0 && stat::minor(self.rdev) == 0
+        }
+        #[cfg(target_os = "macos")]
+        {
+            self.inode.is_chrdev() && ((self.rdev >> 24) & 0xff) == 0 && (self.rdev & 0xffffff) == 0
+        }
     }
 
     /// Check whether the inode (directory) is a overlayfs whiteout opaque.
@@ -1309,29 +1319,29 @@ impl InodeWrapper {
 
     pub fn is_chrdev(&self) -> bool {
         match self {
-            InodeWrapper::V5(i) => i.i_mode & libc::S_IFMT == libc::S_IFCHR,
-            InodeWrapper::V6(i) => i.i_mode & libc::S_IFMT == libc::S_IFCHR,
+            InodeWrapper::V5(i) => i.i_mode & libc::S_IFMT as u32 == libc::S_IFCHR as u32,
+            InodeWrapper::V6(i) => i.i_mode & libc::S_IFMT as u32 == libc::S_IFCHR as u32,
         }
     }
 
     pub fn is_blkdev(&self) -> bool {
         match self {
-            InodeWrapper::V5(i) => i.i_mode & libc::S_IFMT == libc::S_IFBLK,
-            InodeWrapper::V6(i) => i.i_mode & libc::S_IFMT == libc::S_IFBLK,
+            InodeWrapper::V5(i) => i.i_mode & libc::S_IFMT as u32 == libc::S_IFBLK as u32,
+            InodeWrapper::V6(i) => i.i_mode & libc::S_IFMT as u32 == libc::S_IFBLK as u32,
         }
     }
 
     pub fn is_fifo(&self) -> bool {
         match self {
-            InodeWrapper::V5(i) => i.i_mode & libc::S_IFMT == libc::S_IFIFO,
-            InodeWrapper::V6(i) => i.i_mode & libc::S_IFMT == libc::S_IFIFO,
+            InodeWrapper::V5(i) => i.i_mode & libc::S_IFMT as u32 == libc::S_IFIFO as u32,
+            InodeWrapper::V6(i) => i.i_mode & libc::S_IFMT as u32 == libc::S_IFIFO as u32,
         }
     }
 
     pub fn is_sock(&self) -> bool {
         match self {
-            InodeWrapper::V5(i) => i.i_mode & libc::S_IFMT == libc::S_IFSOCK,
-            InodeWrapper::V6(i) => i.i_mode & libc::S_IFMT == libc::S_IFSOCK,
+            InodeWrapper::V5(i) => i.i_mode & libc::S_IFMT as u32 == libc::S_IFSOCK as u32,
+            InodeWrapper::V6(i) => i.i_mode & libc::S_IFMT as u32 == libc::S_IFSOCK as u32,
         }
     }
 

--- a/tests/builder.rs
+++ b/tests/builder.rs
@@ -2,7 +2,8 @@
 //
 // SPDX-License-Identifier: Apache-2.0
 
-use nix::sys::stat::{dev_t, makedev, mknod, Mode, SFlag};
+use nix::sys::stat::{dev_t, mknod, Mode, SFlag};
+use nydus_utils::compact::makedev;
 use std::fs::{self, File};
 use std::io::Write;
 use std::os::unix::fs as unix_fs;

--- a/utils/Cargo.toml
+++ b/utils/Cargo.toml
@@ -20,6 +20,7 @@ serde_json = ">=1.0.9"
 sha2 = "0.10.0"
 tokio = { version = "1.19.0", features = ["rt", "sync"] }
 zstd = "0.11"
+nix = "0.23.1"
 
 nydus-error = { version = "0.2", path = "../error" }
 

--- a/utils/src/compact.rs
+++ b/utils/src/compact.rs
@@ -4,6 +4,8 @@
 // SPDX-License-Identifier: Apache-2.0
 use nix::sys::stat::dev_t;
 
+// makedev calculations compact with macos
+// view https://opensource.apple.com/source/xnu/xnu-201/bsd/sys/types.h.auto.html
 pub fn makedev(major: u64, minor: u64) -> dev_t {
     #[cfg(target_os = "linux")]
     {
@@ -12,5 +14,27 @@ pub fn makedev(major: u64, minor: u64) -> dev_t {
     #[cfg(target_os = "macos")]
     {
         ((major & 0xff << 24) | (minor & 0xffffff)) as dev_t
+    }
+}
+
+pub fn major_dev(dev: u64) -> u64 {
+    #[cfg(target_os = "linux")]
+    {
+        nix::sys::stat::major(dev)
+    }
+    #[cfg(target_os = "macos")]
+    {
+        (dev >> 24) & 0xff
+    }
+}
+
+pub fn minor_dev(dev: u64) -> u64 {
+    #[cfg(target_os = "linux")]
+    {
+        nix::sys::stat::minor(dev)
+    }
+    #[cfg(target_os = "macos")]
+    {
+        dev & 0xffffff
     }
 }

--- a/utils/src/compact.rs
+++ b/utils/src/compact.rs
@@ -1,0 +1,16 @@
+// Copyright 2020 Ant Group. All rights reserved.
+// Copyright (C) 2020 Alibaba Cloud. All rights reserved.
+//
+// SPDX-License-Identifier: Apache-2.0
+use nix::sys::stat::dev_t;
+
+pub fn makedev(major: u64, minor: u64) -> dev_t {
+    #[cfg(target_os = "linux")]
+    {
+        nix::sys::stat::makedev(major, minor)
+    }
+    #[cfg(target_os = "macos")]
+    {
+        ((major & 0xff << 24) | (minor & 0xffffff)) as dev_t
+    }
+}

--- a/utils/src/lib.rs
+++ b/utils/src/lib.rs
@@ -18,6 +18,7 @@ pub use self::inode_bitmap::InodeBitmap;
 pub use self::types::*;
 
 pub mod async_helper;
+pub mod compact;
 pub mod compress;
 pub mod digest;
 pub mod exec;

